sky/cloud_stores.py

```diff
@@ -153,6 +154,135 @@ def make_sync_file_command(self, source: str, destination: str) -> str:
+    # AzCopy for efficient data transfer from containers to local systems.
+    # Note that on Azure instances, both az-cli and AzCopy are typically
+    # pre-installed. And installing both would be used with AZ container is
+    # used from non-Azure instances.
+
+    # Installation reference: https://learn.microsoft.com/en-us/cli/azure/install-azure-cli-linux?pivots=apt#option-2-step-by-step-installation-instructions # pylint: disable=line-too-long
+    _AZ_CLI_VERSION = '2.58.0'
+    _GET_AZCLI = [
+        'az --version >/dev/null 2>&1 || '
+        '(sudo apt-get update; '
+        'sudo apt-get install ca-certificates curl apt-transport-https '
+        'lsb-release gnupg -y; '
+        'sudo mkdir -p /etc/apt/keyrings; '
+        'curl -sLS https://packages.microsoft.com/keys/microsoft.asc | '
+        'gpg --dearmor | '
+        'sudo tee /etc/apt/keyrings/microsoft.gpg > /dev/null; '
+        'sudo chmod go+r /etc/apt/keyrings/microsoft.gpg; '
+        'AZ_DIST=$(lsb_release -cs); '
+        'echo "deb [arch=`dpkg --print-architecture` '
+        'signed-by=/etc/apt/keyrings/microsoft.gpg] '
+        'https://packages.microsoft.com/repos/azure-cli/ $AZ_DIST main" | '
+        'sudo tee /etc/apt/sources.list.d/azure-cli.list; '
+        'sudo apt-get update; '
+        f'AZ_VER={_AZ_CLI_VERSION}; '
+        'sudo apt-get install azure-cli=$AZ_VER-1~$AZ_DIST -y)'
+    ]
+
+    _GET_AZCOPY = [
+        'azcopy --version > /dev/null 2>&1 || '
+        '(mkdir -p /usr/local/bin; '
+        'curl -L https://aka.ms/downloadazcopy-v10-linux -o azcopy.tar.gz; '
+        'sudo tar -xvzf azcopy.tar.gz --strip-components=1 -C /usr/local/bin --exclude=*.txt; '
+        'sudo chmod +x /usr/local/bin/azcopy; '
+        'rm azcopy.tar.gz)'
+    ]
+
+    def is_directory(self, url: str) -> bool:
+        """Returns whether 'url' of the AZ Container is a directory.
+
+        In cloud object stores, a "directory" refers to a regular object whose
+        name is a prefix of other objects.
+        """
+        # split the url using split_az_path
+        container_name, path, storage_account_name = data_utils.split_az_path(
+            url)
+        # If there aren't more than just container name and storage account,
+        # that's a directory.
+        if len(path) == 0:
+            return True
+        # If there's more, we'd need to check if it's a directory or a file.
+        container_client = data_utils.create_az_client(
+            client_type='container',
+            storage_account_name=storage_account_name,
+            container_name=container_name)
+        num_objects = 0
+        for blob in container_client.list_blobs(name_starts_with=path):
+            if blob.name == path:
+                return False
+            num_objects += 1
+            if num_objects > 1:
+                return True
+
+        # A directory with few or no items
+        return True
```

Comment:
`num_objects += 1` will not always let the condition of `if num_objects > 1` to be `True`. It will only be `True` when `num_objects += 1` applies at least twice, which means the given `path` in the `url` is a directory.